### ISSUE TITLE
Fix for #3840 breaks Path encapsulation in PathResource

### ISSUE THREAD

**Issue by jshook** [eclipse/jetty.project#3906]
Fix for Issue #3840 breaks PathResource so that it behaves like a FileResource, and breaks some projects which depend on the newer Path API in the Java runtime. The Path APIs provide useful encpapsulation of filesystems and relatively Paths. 

Using PathResource as such worked fine before, as Jetty was able to find any useful data channel to the Path resource. Now, with a File short-circuiting to the default filesystem, Path resources do not work as intended. Reverting back to 9.4.* fixed the issue for me, but I will not be able to use 10.* unless PathResources continue to work as originally designed.

Simply, using a File instead of a Path causes any relatively-pathed resources to break out of the intended encapsulation back to the default file system, breaking Path abstractions.

This is also possibly a security issue, as file content could be loaded from a file or path outside of a users intended filesystem. This may qualify the issue as a blocker for release of 10.*.

**Comment by sbordet** [comment]
@jshook I think we need more details.
Can you make a clear example of how this change broke others?
Perhaps you can provide a test case?

**Comment by joakime** [comment]
I reopened the old issue #3840 as the fix doesn't seem to improve things on Windows or OSX (in fact on some initial tests on an old windows laptop it seems slower on windows to use File vs Path)
And the fact that byte-range requests seem to use Resource.getInputStream instead of the Resource.getReadableByteChannel is also odd, as Path can return a SeekableByteChannel easily in that case.

**Comment by joakime** [comment]
@sbordet all File objects can be converted to Path objects.
Not all Path objects can be converted to File objects.

An example would be a Path object from a zip-fs or jar-fs.
Or a Path from a in-memory FileSystem. (of which there are few now)
There are also all java FileSystem implementations that provide access to things in a Path / FileSystem syntax (various no-sql databases come to mind here).

**Comment by jshook** [comment]
@sbordet 
By using a FileInputStream which requires a File, you lose one of the primary features of the Path API -- that of filesystem encapsulation. If you have all of your content on the default filesystem, where the File version of a file and the Path version of the file happen to be equivalent in terms of the actual file content to be read, it may not manifest as a bug. Once you try to use a Path on anything besides the default filesystem, the benefit of using a PathResource goes away.

My interpretation for PathResource was that it was originally intended to provide a _Path_ version of the same type of resource as a FileResource, while staying in the safe zone of the Path abstraction.

Paths all know what filesystem they belong to, and Path and Filesystem implementations are intended to be provide a simple and uniform way of navigating relatively or absolutely within a filesystem, thus all buffers, channels ,etc are delegated in some way to the filesystem which a Path is part of. The older File abstraction has no such capability, which is part of the reason that the Path system was derived in the first place.

Before the bugfix, the getInputStream method delegated to the underlying filesystem for the Path in the PathResource. After the bugfix, it allowed a jailbreak by converting the path name to a string and then wrapping it in a File object, which has no concept of what filesystem the original path came from.

**Comment by sbordet** [comment]
@jshook I understand that. 

What are you doing _exactly_ that the change broke? 

I ask so that we can add a concrete test case, as it happened in the past that we "fixed" things based on generic issue reports, only to discover that our "fix" was an improvement but it was not covering that particular use case.

**Comment by jshook** [comment]
Ah, understood. I was using a Path from a non-default filesystem. A virtual filesystem inside the JVM, actually. You can do a test with a Zip filesystem or anything that implements java.nio.file.FileSystem (*not*   java.io.FileSystem) in the Java runtime. Just make sure that the Path you are testing exists within that filesystem, but does not exist on the default "system" filesystem.

Once you use a File to wrap a Path's string representation, any input stream or buffer from it will no longer be attached to the intended resource.

**Comment by joakime** [comment]
I'lll add a testcase for this.

**Comment by joakime** [comment]
Added testcase ...

https://github.com/eclipse/jetty.project/blob/24b2ca4c328fcc22a1c1534d39ea2970920b1ac4/jetty-util/src/test/java/org/eclipse/jetty/util/resource/PathResourceTest.java#L42-L70

Results in the following exception ...

java.lang.UnsupportedOperationException
	at com.sun.nio.zipfs.ZipPath.toFile(ZipPath.java:597)
	at org.eclipse.jetty.util.resource.PathResource.getFile(PathResource.java:368)
	at org.eclipse.jetty.util.resource.PathResource.getInputStream(PathResource.java:384)
	at org.eclipse.jetty.util.resource.PathResourceTest.testNonDefaultFileSystem_GetInputStream(PathResourceTest.java:65)

**Comment by jshook** [comment]
The code I am working on is not easy to drop in for a test case here (It is in active development), so something like the zip test case above is probably the best we could get for now. It covers the case. Thanks for being so quick to cover this!

**Comment by gregw** [comment]
Let's keep discussion in #3840 so it's all in one place

**Comment by joakime** [comment]
Another non-default FileSystem issue encountered has to do with `SeekableByteChannel` from `ZipFileSystem` not supporting ... uh .... seek?

```
2019-07-30 20:20:52.188:WARN:oejs.HttpChannel:qtp1717159510-47: /zipfs/large.dat
java.lang.UnsupportedOperationException
	at com.sun.nio.zipfs.ZipFileSystem$3.position(ZipFileSystem.java:653)
	at org.eclipse.jetty.util.resource.PathResource.writeTo(PathResource.java:591)
	at org.eclipse.jetty.server.ResourceService.sendData(ResourceService.java:755)
	at org.eclipse.jetty.server.ResourceService.doGet(ResourceService.java:294)
	at org.eclipse.jetty.servlet.DefaultServlet.doGet(DefaultServlet.java:454)
```

**Comment by joakime** [comment]
@jshook the fix for this has been merged into `jetty-9.4.x` branch.

There's still some wonkyness with Non-Default FileSystems + Compressed Channels + SeekableByteChannel + Byte Range Requests.
The fact that we can get a UnsupportedOperationException is just heinous. (but understandable with FileSystem's based on compressed content)

**Comment by sbordet** [comment]
> Another non-default FileSystem issue encountered has to do with `SeekableByteChannel` from `ZipFileSystem` not supporting ... uh .... seek?
> 
> ```
> 2019-07-30 20:20:52.188:WARN:oejs.HttpChannel:qtp1717159510-47: /zipfs/large.dat
> java.lang.UnsupportedOperationException
> 	at com.sun.nio.zipfs.ZipFileSystem$3.position(ZipFileSystem.java:653)

I think this is a JDK bug. I'll send an email and see what they reply.
https://mail.openjdk.java.net/pipermail/nio-dev/2019-July/006414.html

**Comment by joakime** [comment]
Java 12.0.1+12 does indeed fix the `ZipFileSystem.position(long)` issue.


Based on the feedback from https://mail.openjdk.java.net/pipermail/nio-dev/2019-July/006415.html

> **From: Florian Weimer (at redhat.com)**
> 
> I think that in practice, FileChannel#position(long) has always been an
> optional operation because it does not work on non-regular files.  This
> isn't specific to zipfs, it can happen with regular file system paths as
> well (if they refer to character devices or FIFOs, for example).

As weird it would be to have those kinds of resources (from Florian's examples) served from a webapp, it does point out that we need a fallback operation.
I'll submit a new PR for a fallback operation for `SeekableByteChannel.position(long)` throwing a `UnsupportedOperationException`

**Comment by joakime** [comment]
Opened PR #3923 for fallback seek mode when `SeekableByteChannel.setPosition(long)` throws an `UnsupportedOperationException`.

**Comment by joakime** [comment]
Fixed in `jetty-9.4.x`
